### ISSUE TITLE
ocamlopt: operand size mismatch for `imul' on 4.12.0

### ISSUE THREAD

**Issue by actionshrimp** [ocaml/ocaml#10626]
I'm seeing the error below from `ocamlopt`.

This is approaching the smallest repro I could find after experimenting with it a bit, removing certain functions inside the module stops the issue from appearing, as does moving everything to top level (outside of a module functor).

The same file runs through `ocamlopt` on 4.08.1 without issue.

```
$ uname -m
x86_64

$ opam exec -- ocamlopt --version
4.12.0

$ cat int64_test.ml
module Datetime = struct
  module Conv (M : sig
    val xs_in_a_second : int64

    val picos_in_an_x : int64
  end) =
  struct
    open Int64

    let second = M.xs_in_a_second

    let minute = mul (of_int 60) second

    let hour = mul (of_int 60) minute

    let day = mul (of_int 24) hour

    let to_picos micros = mul micros M.picos_in_an_x

    let of_picos picos = div picos M.picos_in_an_x

    let to_days micros = div micros day

    let of_days days = mul days day

    let f x =
      let y = to_days x in
      y


    let of_days_picos (days, picos) = add (of_days days) (of_picos picos)
  end
end

$ opam exec -- ocamlopt int64_test.ml
/tmp/camlasm6c2cb6.s: Assembler messages:
/tmp/camlasm6c2cb6.s:51: Error: operand size mismatch for `imul'
File "int64_test.ml", line 1:
Error: Assembler error, input left in file /tmp/camlasm6c2cb6.s
```

**Comment by gasche** [comment]
Reproduction note: I can reproduce the issue (amd64 backend, without flambda), and it seems to be new in 4.12.0, the repro cas works with 4.11.1.

**Comment by nojb** [comment]
Culprit is b735f187acaf284148f0e41ca86be8eb08347230 cc @stedolan

**Comment by lthls** [comment]
The bug is that we generate the following assembler instruction:
```
	imulq	$60, (%rsp)
```
This is illegal (multiplications need to have their result and the corresponding argument in a register), but before Stephen's commit it would never occur: either the multiplication was done on OCaml integers, and the result of the multiplication never needed to be stack-allocated, or the multiplication was done on boxed integers, and the operation would use a regular two-argument multiplication instead of a single-argument multiplication by a constant, which is handled correctly.
I believe the fix is to treat `Iintop_imm(Imul, _)` as `Iintop (Imul)` in `reload_operation` (in `amd64/reload.ml`), forcing a hard register for the argument; but I don't know if there are other operations or platforms that need to be fixed too.

For reference, here is a smaller example that also triggers the bug:
```ocaml
(* This module is only needed with flambda, as the functions
   don't get inlined with the default flambda flags. *)
module Int64 = struct
  let[@inline always] add x y = (Int64.add[@inlined always]) x y
  let[@inline always] mul x y = (Int64.mul[@inlined always]) x y
end

let f n =
  let n1 = Int64.add n 1L in
  let n2 = Int64.mul 60L n1 in
  let n3 = Int64.add n2 2L in
  let n4 = Int64.add n3 3L in
  let n5 = Int64.add n4 4L in
  let n6 = Int64.add n5 5L in
  let n7 = Int64.add n6 6L in
  let n8 = Int64.add n7 7L in
  let n9 = Int64.add n8 8L in
  (n1, n2, n3, n4, n5, n6, n7, n8, n9)
```

**Comment by xavierleroy** [comment]
@lthls 's analysis is correct: this is a bug in the amd64/reload.ml file.  Interestingly, i386/reload.ml gets it right.  Here is the trivial fix: #10628 .
